cns/restserver/metrics.go

```diff
@@ -142,8 +144,51 @@ func stateTransitionMiddleware(i *cns.IPConfigurationStatus, s types.IPState) {
 	ipConfigStatusStateTransitionTime.WithLabelValues(string(i.GetState()), string(s)).Observe(time.Since(i.LastStateTransition).Seconds())
 }
 
-func publishIPStateMetrics(state *ipState) {
-	labels := []string{} // TODO. ragasthya Add dimensions to the IP Usage metrics.
+type ipState struct {
+	// allocatedIPs are all the IPs given to CNS by DNC.
+	allocatedIPs int64
+	// assignedIPs are the IPs CNS gives to Pods.
+	assignedIPs int64
+	// availableIPs are the IPs in state "Available".
+	availableIPs int64
+	// programmingIPs are the IPs in state "PendingProgramming".
+	programmingIPs int64
+	// releasingIPs are the IPs in state "PendingReleasr".
+	releasingIPs int64
+}
+
+// publishIPStateMetrics logs and publishes the IP Config state metrics to Prometheus.
+func (service *HTTPRestService) publishIPStateMetrics() {
+	// copy state
+	service.RLock()
```

Comment:
Could this cause a slow hangup from the server if say:

G1: Finishes handler
G1: defer service.Unlock() runs
G1: publishIPStateMetrics() starts
G2: Starts handler
G2: service.Lock()
G2: sleep for an eternity
G2: defer service.Unlock() runs
G2: publishIPStateMetrics() runs to completion
G1: service.RLock()
G1: finishes publishIPStateMetrics() and releases locks
G1: writes 200 and closes socket now that the handler func is finally done

Reply:
valid point, need to reconsider this. i hate just shoving more things in to the server Locked scope but the mutexing on this god object is not very precise